packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,266 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2020 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""HTTP connection and channel."""
+
+import asyncio
+import logging
+import threading
+import time
+import json
+import yaml
+from asyncio import CancelledError
+from threading import Thread
+from typing import Any, Dict, List, Optional, Set, cast
+
+from fetch.p2p.api.http_calls import HTTPCalls
+
+from aea.configurations.base import ConnectionConfig, PublicId
+from aea.connections.base import Connection
+from aea.helpers.base import locate
+from aea.mail.base import AEAConnectionError, Address, Envelope, EnvelopeContext
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPChannel:
+    """A wrapper for an RESTful API."""
+
+    def __init__(
+        self,
+        address: Address,
+        api_spec: str,
+        provider_addr: str,
+        provider_port: int,
+        excluded_protocols: Optional[Set[PublicId]] = None,
+    ):
+        """
+        Initialize a channel.
+
+        :param address: the address
+        """
+        self.address = address
+        self.provider_addr = provider_addr
+        self.provider_port = provider_port
+        self.in_queue = None  # type: Optional[asyncio.Queue]
+        self.loop = None  # type: Optional[asyncio.AbstractEventLoop]
+        self._httpCall = None  # type: Optional[HTTPCalls]
+        self.excluded_protocols = excluded_protocols
+        self.thread = Thread(target=self.receiving_loop)
+        self.lock = threading.Lock()
+        self.stopped = True
+        logger.info("Initialised the http channel")
+
+    def connect(self):
+        """
+        Connect.
+
+        :return: an asynchronous queue, that constitutes the communication channel.
+        """
+        with self.lock:
+            if self.stopped:
+                self._httpCall = HTTPCalls(
```

Comment:
not relevant, as this is supposed to be the server

Reply:
to start with, you could make HTTPChannel inherit from https://docs.python.org/2/library/basehttpserver.html#BaseHTTPServer.BaseHTTPRequestHandler
and here an example https://wiki.python.org/moin/BaseHttpServer. Then you just need to implement the `do_..` methods. i.e. do_GET, do_HEAD, do POST and create the envelopes which you hand off to the Connection

Reply:
this isn't ready for production but it is ok for the demo

Reply:
Thanks for that, I will check out BaseHTTPServer.